transactions/src/main/java/org/jboss/as/txn/service/internal/tsr/JCAOrderedLastSynchronizationList.java

```diff
@@ -0,0 +1,121 @@
+package org.jboss.as.txn.service.internal.tsr;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import javax.transaction.Synchronization;
+import javax.transaction.Transaction;
+
+import org.jboss.as.txn.logging.TransactionLogger;
+
+/**
+ * This class was added to:
+ *
+ * 1. workaround an issue discussed in https://java.net/jira/browse/JTA_SPEC-4 whereby the JCA Synchronization(s) need to be
+ * called after the JPA Synchronization(s). Currently the implementation orders JCA relative to all interposed Synchronizations,
+ * if this is not desirable it would be possible to modify this class to store just the JPA and JCA syncs and the other syncs
+ * can simply be passed to a delegate (would need the reference to this in the constructor).
+ *
+ * 2. During afterCompletion the JCA synchronizations should be called last as that allows JCA to detect connection leaks from
+ * frameworks that have not closed the JCA managed resources. This is described in (for example)
+ * http://docs.oracle.com/javaee/5/api/javax/transaction/TransactionSynchronizationRegistry
+ * .html#registerInterposedSynchronization(javax.transaction.Synchronization) where it says that during afterCompletion
+ * "Resources can be closed but no transactional work can be performed with them"
+ */
+public class JCAOrderedLastSynchronizationList implements Synchronization {
+    private Transaction tx;
+    private Map<Transaction, JCAOrderedLastSynchronizationList> jcaOrderedLastSynchronizations;
+    private List<Synchronization> preJcaSyncs = new ArrayList<Synchronization>();
+    private List<Synchronization> jcaSyncs = new ArrayList<Synchronization>();
+
+    public JCAOrderedLastSynchronizationList(Transaction tx,
+        Map<Transaction, JCAOrderedLastSynchronizationList> jcaOrderedLastSynchronizations) {
+        this.tx = tx;
+        this.jcaOrderedLastSynchronizations = jcaOrderedLastSynchronizations;
+        this.jcaOrderedLastSynchronizations.put(tx, this);
```

Comment:
Ah, this probably relates to the putIfAbsent comment. See my response there as I am not sure I can use putIfAbsent as if it is absent I need to also register it with the delegate TSR and I don't see a callback in the putIfAbsent call where I could do that.